AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -218,309 +232,137 @@ - (void)_batchLayoutAndInsertNodesFromContexts:(NSArray<ASIndexedNodeContext *>
-  }];
+  return indexPaths;
 }
 
 /**
- * Removes the specified nodes at the given index paths and notifies the delegate of the nodes removed.
+ * Agressively repopulates supplementary nodes of all kinds for sections that contains some given index paths.
  *
- * @discussion Nodes are first removed from the editing store then removed from the completed store on the main thread.
- * Once the backing stores are consistent, the delegate is invoked on the main thread.
+ * @param originalIndexPaths The index paths belongs to sections whose supplementary nodes need to be repopulated.
+ * @param environment The trait environment needed to initialize contexts
  */
-- (void)_deleteNodesAtIndexPaths:(NSArray *)indexPaths withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions
+- (void)_repopulateSupplementaryNodesForAllSectionsContainingIndexPaths:(NSArray<NSIndexPath *> *)originalIndexPaths
+                                                            environment:(id<ASTraitEnvironment>)environment
 {
-  ASSERT_ON_EDITING_QUEUE;
+  ASDisplayNodeAssertMainThread();
   
-  [self deleteNodesOfKind:ASDataControllerRowNodeKind atIndexPaths:indexPaths completion:^(NSArray *nodes, NSArray *indexPaths) {
-    ASDisplayNodeAssertMainThread();
-    
-    [_delegate dataController:self didDeleteItemsAtIndexPaths:indexPaths withAnimationOptions:animationOptions];
-  }];
-}
-
-/**
- * Inserts sections, represented as arrays, into the backing store at the given indices and notifies the delegate.
- *
- * @discussion The section arrays are inserted into the editing store, then a deep copy of the sections are inserted
- * in the completed store on the main thread. The delegate is invoked on the main thread.
- */
-- (void)_insertSections:(NSMutableArray *)sections atIndexSet:(NSIndexSet *)indexSet withAnimationOptions:(ASDataControllerAnimationOptions)animationOptions
-{
-  ASSERT_ON_EDITING_QUEUE;
+  if (originalIndexPaths.count ==  0) {
+    return;
+  }
   
-  [self insertSections:sections ofKind:ASDataControllerRowNodeKind atIndexSet:indexSet completion:^(NSArray *sections, NSIndexSet *indexSet) {
-    ASDisplayNodeAssertMainThread();
-    
-    [_delegate dataController:self didInsertSections:sections atIndexSet:indexSet withAnimationOptions:animationOptions];
-  }];
+  // Get all the sections that need to be repopulated
+  NSIndexSet *sectionIndexes = [NSIndexSet as_sectionsFromIndexPaths:originalIndexPaths];
+  for (NSString *kind in [self supplementaryKindsInSections:sectionIndexes]) {
+    // Step 1: Remove all existing contexts of this kind in these sections
+    // TODO: NSEnumerationConcurrent?
+    // TODO: Consider using a diffing algorithm here
+    [_nodeContexts[kind] enumerateObjectsAtIndexes:sectionIndexes options:0 usingBlock:^(NSMutableArray<ASIndexedNodeContext *> * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+      [obj removeAllObjects];
+    }];
```

Comment:
Would it make more sense to do `_nodeContexts enumerateKeysAndObjectsUsingBlock:` for this removal step? That way we are sure we removed all the old supplementaries, even if that kind isn't present anymore.